File: src/java/model/Joust.java
package model;

import java.util.Arrays;
import java.util.stream.Stream;

public class Joust {

    private final int rows;
    private final int cols;
    private final CellState[][] board;
    private Player turn = Player.PLAYER1;
    private MoveResult gameIsOver = new MoveResult(Move.VALID);

    public Joust(int nrows, int ncols) {
        this.rows = nrows;
        this.cols = ncols;
        this.board = this.startBoard();
    }

    private CellState[][] startBoard() {
        CellState[][] matrix = new CellState[rows][cols];
        for (CellState[] row : matrix) {
            Arrays.fill(row, CellState.EMPTY);
        }
        CellState[] players = {CellState.PLAYER1, CellState.PLAYER2};
        for (int i = 0; i < players.length;) {
            int row = (int) Math.floor(Math.random() * rows);
            int col = (int) Math.floor(Math.random() * cols);
            if (matrix[row][col] == CellState.EMPTY) {
                matrix[row][col] = players[i];
                i++;
            }
        }
        return matrix;
    }

    public MoveResult move(Player player, Cell endCell) {
        if (gameIsOver.getWinner() != null) {
            return gameIsOver;
        }
        if (player != turn) {
            return new MoveResult(Move.INVALID);
        }
        Cell beginCell = getPlayerCell(player);
        MoveResult m = isValidMove(beginCell, endCell);
        if (m.getMove() == Move.VALID) {
            int or = beginCell.getX(), oc = beginCell.getY();
            int dr = endCell.getX(), dc = endCell.getY();
            board[dr][dc] = board[or][oc];
            board[or][oc] = CellState.BLOCKED;
            turn = (turn == Player.PLAYER1) ? Player.PLAYER2 : Player.PLAYER1;
            gameIsOver = isGameOver();
            return gameIsOver;
        }
        return m;
    }

    private MoveResult isValidMove(Cell beginCell, Cell endCell) {
        int or = beginCell.getX(), oc = beginCell.getY();
        int dr = endCell.getX(), dc = endCell.getY();
        /* Destino deve estar vazio */
        if (board[dr][dc] != CellState.EMPTY) {
            return new MoveResult(Move.INVALID);
        }
        int[] lin = {-2, 2, -1, 1};
        int[][] inc = {{-1, 1}, {-1, 1}, {-2, 2}, {-2, 2}};
        for (int i = 0; i < lin.length; i++) {
            for (int j = 0; j < inc[i].length; j++) {
                Cell cell = new Cell(or + lin[i], oc + inc[i][j]);
                if (isValidCell(cell) && cell.equals(endCell)) {
                    return new MoveResult(Move.VALID);
                }
            }
        }
        return new MoveResult(Move.INVALID);
    }

    private Cell getPlayerCell(Player player) {
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {
                if ((board[i][j] == CellState.PLAYER1 && player == Player.PLAYER1)
                        || (board[i][j] == CellState.PLAYER2 && player == Player.PLAYER2)) {
                    return new Cell(i, j);
                }
            }
        }
        return null;
    }

    private boolean canMove(Player player) {
        Cell cell = getPlayerCell(player);
        boolean ok = false;
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (isValidMove(cell, new Cell(i, j)).getMove() == Move.VALID) {
                    ok = true;
                }
            }
        }
        return ok;
    }

    private MoveResult isGameOver() {
        boolean p1 = canMove(Player.PLAYER1);
        boolean p2 = canMove(Player.PLAYER2);
        MoveResult mr = new MoveResult(Move.VALID);
        if (!p1 && !p2) {
            mr.setWinner(Winner.DRAW);
        } else if (!p1) {
            mr.setWinner(Winner.PLAYER2);
        } else if (!p2) {
            mr.setWinner(Winner.PLAYER1);
        }
        return mr;
    }

    public Winner getWinner() {
        return isGameOver().getWinner();
    }

    private boolean isValidCell(Cell cell) {
        return (cell.getX() < rows && cell.getX() >= 0 && cell.getY() < cols && cell.getY() >= 0);
    }

    public Player getTurn() {
        return turn;
    }

    public CellState[][] getBoard() {
        return board;
    }

    public static void main(String[] args) {
        Joust j = new Joust(8, 8);
        Stream.of(j.getBoard()).forEach(x -> System.out.println(Arrays.toString(x)));
    }
}

File: docs/Joust.js
import {CellState} from './CellState.js';
import {Player} from './Player.js';

function Joust(nrows, ncols) {
    const rows = nrows;
    const cols = ncols;
    let board = startBoard();
    let turn = Player.PLAYER1;
    function getBoard() {
        return board;
    }
    function startBoard() {
        let matrix = Array(rows).fill().map(() => Array(cols).fill(CellState.EMPTY));
        let players = [CellState.PLAYER1, CellState.PLAYER2];
        for (let i = 0; i < players.length; ) {
            let row = Math.floor(Math.random() * rows);
            let col = Math.floor(Math.random() * cols);
            if (matrix[row][col] === CellState.EMPTY) {
                matrix[row][col] = players[i++];
            }
        }
        return matrix;
    }
    function getTurn() {
        return turn;
    }
    function move(beginCell, endCell) {
        let {x: or, y: oc} = beginCell;
        let {x: dr, y: dc} = endCell;
        if (!beginCell || !endCell) {
            throw new Error("The value of one of the cells does not exist.");
        }
        let currentPiece = getPiece(beginCell);
        if ((currentPiece === CellState.PLAYER1 && turn === Player.PLAYER2) || (currentPiece === CellState.PLAYER2 && turn === Player.PLAYER1)) {
            throw new Error("It's not your turn.");
        }
        if (beginCell.equals(endCell)) {
            throw new Error("Origin and destination must be different.");
        }
        if (!onBoard(beginCell) || !onBoard(endCell)) {
            throw new Error("Origin or destination are not in the board.");
        }
        if (currentPiece === CellState.EMPTY) {
            throw new Error("Origin does not have a piece.");
        }
        if (getPiece(endCell) !== CellState.EMPTY) {
            throw new Error("Destination must be empty.");
        }
        /* Realizar movimento */
        board[dr][dc] = board[or][oc];
        board[or][oc] = CellState.BLOCKED;
        turn = (turn === Player.PLAYER1) ? Player.PLAYER2 : Player.PLAYER1;
    }
    function getPiece( {x, y}) {
        return board[x][y];
    }
    function onBoard( {x, y}) {
        let inLimit = (value, limit) => value >= 0 && value < limit;
        return (inLimit(x, rows) && inLimit(y, cols));
    }
    return {getBoard, getTurn, move};
}

export {Joust};